risc0/zkp/verify/taps.h

```diff
@@ -0,0 +1,259 @@
+// Copyright 2022 Risc0, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <map>
+#include <set>
+#include <vector>
+
+namespace risc0 {
+
+// The tap set specifies for each register group and for each register, which
+// timestaps back we 'tap'.  We group taps at two levels, first by register and
+// then by register group.  Additionally, multiple registers (within or across
+// groups) which have the same pattern of taps are part of the same 'combo'.
+// These taps may be accumulated together in some places for later division by
+// the same value.
+
+enum class RegisterGroup : uint8_t {
+  ACCUM = 0,
+  CODE = 1,
+  DATA = 2,
+  INVALID = 3,
+};
+
+static constexpr size_t kNumRegisterGroups = 3;
+
+namespace impl {
```

Comment:
nit: Idiomatic? c++ code usually uses `detail`. Also thinking that I might want to use `impl` as var name more often then I'd use `detail` as a var name.